src/System.Private.CoreLib/shared/System/Diagnostics/Debug.cs

```diff
@@ -9,101 +9,138 @@
 
 namespace System.Diagnostics
 {
+    public abstract class TraceDebugBase
+    {
+        public abstract bool AutoFlush { get; set; }
+        public abstract void Assert(bool condition);
+        public abstract void Assert(bool condition, string message);
+        public abstract void Assert(bool condition, string message, string detailMessage);
+        public abstract void Close();
+        public abstract void Fail(string message);
+        public abstract void Fail(string message, string detailMessage);
+        public abstract void Flush();
+        public abstract int IndentLevel { get; set; }
+        public abstract int IndentSize { get; set; }
+        public abstract void Indent();
+        public abstract void Unindent();
+        public abstract void Write(object value);
+        public abstract void Write(object value, string category);
+        public abstract void Write(string message);
+        public abstract void Write(string message, string category);
+        public abstract void WriteLine(object value);
+        public abstract void WriteLine(object value, string category);
+        public abstract void WriteLine(string message);
+        public abstract void WriteLine(string message, string category);
+    }
+
     /// <summary>
     /// Provides a set of properties and methods for debugging code.
     /// </summary>
     public static partial class Debug
     {
-        private static readonly object s_lock = new object();
+        private static TraceDebugBase _impl;
+        private static TraceDebugBase Impl
+        {
+            get
+            {
+                if (_impl == null)
+                {
+                    TraceDebugBase debugImplementation = null;
+                    Type typeFromTraceSource = null;
+                    try
+                    {
+                        // typeFromTraceSource = Type.GetType(
```

Comment:
I get StackOverflowException even when I uncomment only lines 52 to 55

Reply:
I'll try a third approach instead to continue on with this prototype and that involves making reflection call on corefx instead of here.